### ISSUE TITLE
Nodes in test dataset but not in training dataset

### ISSUE THREAD

**Issue by chad-sharethis** [facebookresearch/PyTorch-BigGraph#174]
PBG splits the data into training and test datasets and trains on the training dataset. When we evaluate the performance using the test dataset, are we assuming all the nodes in the test dataset are also present in the training dataset? If not, what embedding values will be assigned to those nodes present in test dataset only?

**Comment by chad-sharethis** [comment]
Thanks.
Is it possible when it does the train/test split, some nodes in the test may not be contained in the training set, and therefore, no embeddings are available for these nodes?

**Comment by lw** [comment]
It could happen that all the edges incident to a certain node end up in the test set, and thus in the train set the node is "isolated" and thus its embedding will end up being random noise. However, the pre-processor that we ship with PBG does not do the train/test split on its own (as far as I remember) but instead expects the input data to already be split. If you're splitting it yourself, that might be the issue. Note however that typically the test is a very small fraction of the train (like, 5%?), so if all the edges of a node end up in the test split either you've been incredibly unlucky or the node didn't have many edges to start with. A node with low degree won't probably learn a good embedding anyways, so we suggest to just prune it, and if you really need an embedding you can obtain a pretty good proxy for it by taking the average embedding of its neighbors.

**Comment by chad-sharethis** [comment]
Thanks again for the explanation.
For FB15k, it's true that all the nodes in test.txt is also in train.txt.

On the other hand, in the livejournal example, the data is split into train.txt and test.txt by the ratio of 75/25. I do see quite some ids that are in test.txt but are not in train.txt. My understanding is that during the training those ids won't show up in positive pairs but can still get pulled in by uniform negative sampling and show up in the negative pairs.

How will this affect the metrics of the evaluation job run at the end of the livejournal.py script?
